pkg/cloudprovider/gcp.go

```diff
@@ -1,47 +1,209 @@
+func (p *GCPProvider) UpdateRegistryDC(dc *appsv1.DeploymentConfig, name, dirName string) {
+	dc.Spec.Template.Spec.Containers[0].Env = []kapi.EnvVar{
+		{
+			Name:  "REGISTRY_STORAGE",
+			Value: "gcs",
+		},
+		{
+			Name:  "REGISTRY_STORAGE_GCS_BUCKET",
+			Value: p.Bucket,
+		},
+		{
+			Name:  "REGISTRY_STORAGE_GCS_ROOTDIRECTORY",
+			Value: "/" + dirName,
+		},
+		{
+			Name:  "REGISTRY_STORAGE_GCS_KEYFILE",
+			Value: "/credentials/cloud",
+		},
+	}
+	dc.Spec.Template.Spec.Containers[0].VolumeMounts = []kapi.VolumeMount{
+		{
+			Name:      "cloud-credentials",
+			MountPath: "/credentials",
+		},
+	}
+	dc.Spec.Template.Spec.Volumes = []kapi.Volume{
+		{
+			Name: "cloud-credentials",
+			VolumeSource: kapi.VolumeSource{
+				Secret: &kapi.SecretVolumeSource{
+					SecretName: name,
+				},
+			},
+		},
+	}
 }
 
 func (p *GCPProvider) Validate(secret *kapi.Secret) []string {
 	fields := []string{}
 
+	if secret != nil {
+		keySet := []string{
+			GcpCredentials,
+		}
+		for _, k := range keySet {
+			p, _ := secret.Data[k]
+			if p == nil || len(p) == 0 {
+				fields = append(fields, "Secret(content)")
+				break
+			}
+		}
+	}
+
 	switch p.Role {
 	case BackupStorage:
+		if p.Bucket == "" {
+			fields = append(fields, "Bucket")
+		}
 	case VolumeSnapshot:
 	}
 
 	return fields
 }
 
 func (p *GCPProvider) Test(secret *kapi.Secret) error {
+	var err error
 	if secret == nil {
 		return nil
 	}
 
+	switch p.Role {
+	case BackupStorage:
+		key, _ := uuid.NewUUID()
+		test := GcsTest{
+			key:    key.String(),
+			bucket: p.Bucket,
+			secret: secret,
+		}
+		err = test.Run()
+		if err != nil {
+			return err
+		}
+	case VolumeSnapshot:
+		// TBD
+	}
+
 	return nil
 }
+
+type GcsTest struct {
+	bucket string
+	secret *kapi.Secret
+	key    string
+}
+
+func (r *GcsTest) Run() error {
+	client, err := r.newClient()
+	if err != nil {
+		return err
+	}
+	defer client.Close()
+	err = r.upload(client)
+	if err != nil {
+		return err
+	}
+	defer r.delete(client)
+	err = r.download(client)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (r *GcsTest) newClient() (*storage.Client, error) {
+	client, err := storage.NewClient(
+		context.Background(),
+		option.WithScopes(storage.ScopeReadWrite),
+		option.WithCredentialsJSON(r.secret.Data[GcpCredentials]))
+	if err != nil {
+		return nil, err
+	}
+
+	return client, nil
+}
+
+func (r *GcsTest) upload(client *storage.Client) error {
+	bucket := client.Bucket(r.bucket)
+	object := bucket.Object(r.key)
+	writer := object.NewWriter(context.Background())
+	_, err := writer.Write([]byte{0})
+	if err != nil {
+		writer.Close()
+		return err
+	}
+	err = writer.Close()
+	return err
+}
+
+func (r *GcsTest) download(client *storage.Client) error {
+	bucket := client.Bucket(r.bucket)
+	object := bucket.Object(r.key)
+	reader, err := object.NewReader(context.Background())
+	if err != nil {
+		return err
+	}
+	_, err = reader.Read(make([]byte, 1))
+	reader.Close()
```

Comment:
I don't know how likely it is for reader.Close() to return an error, but we should probably check the return value here like we're doing with writer.Close() since it's possible for reader.Close() to fail.

Reply:
Done.